developer-hub/src/reference_docs/safe_defensive_programming.rs

```diff
@@ -1 +1,376 @@
+//! return `Option<T>`.
+//!
+//! #### Handling via Result - Less Verbose
+//!
+//! In the Polkadot SDK codebase, you may see checked operations being handled as a [`Result`] via
+//! `ok_or`. This is a less verbose way of expressing the above.  Which to use often boils down to
+//! the developer's preference:
+//!
+#![doc = docify::embed!(
+    "./src/reference_docs/safe_defensive_programming.rs",
+    increase_balance_result
+)]
+//!
+//!
+//! ### Saturated Operations
+//!
+//! Saturating a number limits it to the type's upper or lower bound, no matter the integer would
+//! overflow in runtime. For example, adding to `u32::MAX` would simply limit itself to `u32::MAX`:
+//!
+#![doc = docify::embed!(
+    "./src/reference_docs/safe_defensive_programming.rs",
+    saturated_add_example
+)]
+//!
+//! Saturating calculations can be used if one is very sure that something won't overflow, but wants to
+//! avoid introducing the notion of any potential-panic or wrapping behavior.
+//!
+//! ### Mathematical Operations in Substrate Development - Further Context
+//!
+//! As a recap, we covered the following concepts:
+//!
+//! 1. **Checked** operations - using [`Option`] or [`Result`],
+//! 2. **Saturated** operations - limited to the lower and upper bounds of a number type,
+//! 3. **Wrapped** operations (the default) - wrap around to above or below the bounds of a type,
+//!
+//! #### The problem with 'default' wrapped operations
+//!
+//! **Wrapped operations** cause the overflow to wrap around to either the maximum or minimum of that
+//! type. Imagine this in the context of a blockchain, where there are balances, voting counters, nonces
+//! for transactions, and other aspects of a blockchain.
+//!
+//! Some of these mechanisms can be more critical than others. It's for this reason that we may consider
+//! some other ways of dealing with runtime arithmetic, such as saturated or checked operations, that
+//! won't carry these potential consequences.
+//!
+//!
+//! While it may seem trivial, choosing how to handle numbers is quite important. As a thought exercise,
+//! here are some scenarios of which will shed more light on when to use which.
+//!
+//! #### Bob's Overflowed Balance
+//!
+//! **Bob's** balance exceeds the `Balance` type on the `EduChain`. Because the pallet developer did not
+//! handle the calculation to add to Bob's balance with any regard to this overflow, **Bob's** balance
+//! is now essentially `0`, the operation **wrapped**.
+//!
+//! <details>
+//!     <summary><b>Solution: Saturating or Checked</b></summary>
+//!     For Bob's balance problems, using a `saturated_add` or `checked_add` could've mitigated this issue.  They simply would've reached the upper, or lower bounds, of the particular type for an on-chain balance.  In other words: Bob's balance would've stayed at the maximum of the Balance type.
+//! </details>
+//!
+//! #### Alice's 'Underflowed' Balance
```

Comment:
Maybe you can also mention issues with Div/Mod by zero? It is equally problematic to use `/` or `%` in such contexts.

Reply:
yes good point, will add, thank you! good opportunity to stick `checked_div` in there